src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/ModelBindingContext.cs

```diff
@@ -27,60 +22,75 @@ public ModelBindingContext()
         /// <param name="metadata"><see cref="ModelMetadata"/> associated with the model.</param>
         /// <param name="bindingInfo"><see cref="BindingInfo"/> associated with the model.</param>
-        /// <param name="modelName">An optional name of the model to be used.</param>
+        /// <param name="modelName">The name of the property or parameter being bound.</param>
         /// <returns>A new instance of <see cref="ModelBindingContext"/>.</returns>
-        public static ModelBindingContext GetModelBindingContext(
+        public static ModelBindingContext CreateBindingContext(
+            [NotNull] OperationBindingContext operationBindingContext,
+            [NotNull] ModelStateDictionary modelState,
             [NotNull] ModelMetadata metadata,
             BindingInfo bindingInfo,
-            string modelName)
+            [NotNull] string modelName)
         {
             var binderModelName = bindingInfo?.BinderModelName ?? metadata.BinderModelName;
             var propertyPredicateProvider =
                 bindingInfo?.PropertyBindingPredicateProvider ?? metadata.PropertyBindingPredicateProvider;
-            return new ModelBindingContext
+
+            return new ModelBindingContext()
             {
-                ModelMetadata = metadata,
+                BinderModelName = binderModelName,
                 BindingSource = bindingInfo?.BindingSource ?? metadata.BindingSource,
-                PropertyFilter = propertyPredicateProvider?.PropertyFilter,
                 BinderType = bindingInfo?.BinderType ?? metadata.BinderType,
-                BinderModelName = binderModelName,
-                ModelName = binderModelName ?? metadata.PropertyName ?? modelName,
+                PropertyFilter = propertyPredicateProvider?.PropertyFilter,
+
+                // We only support fallback to empty prefix in cases where the model name is inferred from
+                // the parameter or property being bound.
                 FallbackToEmptyPrefix = binderModelName == null,
+
+                // Because this is the top-level context, FieldName and ModelName should be the same.
+                FieldName = binderModelName ?? modelName,
+                ModelName = binderModelName ?? modelName,
+
+                IsTopLevelObject = true,
+                ModelMetadata = metadata,
+                ModelState = modelState,
+                OperationBindingContext = operationBindingContext,
+                ValueProvider = operationBindingContext.ValueProvider,
+            };
+        }
+
+        public static ModelBindingContext CreateChildBindingContext(
+            [NotNull] ModelBindingContext parent,
+            [NotNull] ModelMetadata modelMetadata,
+            [NotNull] string fieldName,
+            [NotNull] string modelName,
+            object model)
+        {
+            return new ModelBindingContext()
+            {
+                IsFirstChanceBinding = parent.IsFirstChanceBinding,
```

Comment:
This changes the meaning of `IsFirstChanceBinding`. See `<remarks>` on that property: `Should not be <c>true</c> when <see cref="IsTopLevelObject"/> is <c>false</c>.`